### ISSUE TITLE
Partial outputs

### ISSUE THREAD

**Issue by MisterTea** [fchollet/keras#1516]
Hey there!

I have a network where only one of the outputs is known for any training example.  How can I prevent the other outputs from propagating an error?  Is there some sentinel value I can set them to, or some way to customize the loss function to ignore all but one output node?

Thanks!

**Comment by MisterTea** [comment]
Actually I think I have the objective function that does this (assuming the undefined outputs are set to 0):

```
def non_zero_mean_squared_error(y_true, y_pred):
    return theano.tensor.mean(theano.tensor.square(theano.tensor.sgn(y_true)*(y_pred - y_true)), axis=-1)
```

Does this look reasonable?  The idea is that only one of the true label outputs is non-zero (and always positive) so the sgn function will allow the non-zero outputs to contribute to the error and the rest have no effect.

**Comment by lemuriandezapada** [comment]
Why not just assign some arbitrarily unlucky value to confirm the 0? Such as let's say 65535 or something.

**Comment by MisterTea** [comment]
I don't understand your comment?  The issue is that the training examples do not have a value for all of the output nodes. If I put a sentinel value (even 0 or 65535) and use a normal loss function, it will train the network to learn these sentinel values. What I need is a loss function that purposely ignores some of the outputs. I believe the one I wrote above will do this.

**Comment by MisterTea** [comment]
Actually the function I wrote doesn't work, because the "mean" function will consider the 0's in the list when dividing.

**Comment by MisterTea** [comment]
This function should work correctly:

```
def non_zero_mean_squared_error(y_true, y_pred):
    total = theano.tensor.sum(theano.tensor.square(theano.tensor.sgn(y_true)*(y_pred - y_true)), axis=-1)
    count = theano.tensor.sum(theano.tensor.sgn(y_true))
    return total/count
```

**Comment by lemuriandezapada** [comment]
I was thinking more like:
theano.tensor.mean(theano.tensor.square((y_true==sentinel_value)*(y_pred - y_true)), axis=-1)

That way it should also work for regression.
Does the "mean" actually matter for the loss function's behavior? In the end it's only some scalar.

**Comment by MisterTea** [comment]
The size of the gradient affects the learning rate, so yes it is important to get the mean correct.

Your point about using eq instead of sgn makes sense.  Here's a function that uses eq instead:

```
def non_zero_mean_squared_error(y_true, y_pred):
    total = theano.tensor.sum(theano.tensor.square(theano.tensor.eq(y_true,0)*(y_pred - y_true)), axis=-1)
    count = theano.tensor.sum(theano.tensor.eq(y_true,0))
    return total/count
```
